src/decoder.rs

```diff
@@ -0,0 +1,181 @@
+//! Parse ITM packets from bytes and streams.
+
+use error::{Error, ErrorKind, Result};
+use heapless::Vec as HVec;
+use packet::{self, Packet, Instrumentation};
+use std::io::{self, Cursor, Read};
+
+/// Parses ITM packets.
+pub struct Decoder<R: Read> {
+    inner: R,
+}
+
+impl<R: Read> Decoder<R> {
+    /// Construct a new `Decoder` that reads encoded packets from
+    /// `inner`.
+    pub fn new(inner: R) -> Decoder<R> {
+        Decoder::<R> {
+            inner: inner,
+        }
+    }
+
+    // TODO: If we need config for the Decoder, my plan is to:
+    // * Add a Config struct with private fields that can be used with
+    //   `serde`, built with a builder pattern (`derive_builder` is
+    //   great), or built with a Default implementation.
+    // * Add a method Decoder.with_config(inner: R, config: Config).
+
+    /// Read a single packet from the inner `Read`. This will block
+    /// for input if no full packet is currently an available.
+    pub fn read_packet(&mut self) -> Result<Packet> {
+        let mut header = [0; 1];
+        match self.inner.read_exact(&mut header) {
+            Err(ref e) if e.kind() == io::ErrorKind::UnexpectedEof =>
+                return Err(Error::from(ErrorKind::EofBeforePacket)),
+            Err(e) => return Err(Error::from(e)),
+            Ok(_) => (),
+        };
+        let header = header[0];
+        match header & 0b111 {
+            0b001|0b010|0b011 => {
+                // Instrumentation packet.
+                let mut ud = Instrumentation {
+                    payload: HVec::new(),
+                    port: header >> 3,
+                };
+
+                let payload_size =
+                    match header & 0b11 {
+                        0b01 => 1,
+                        0b10 => 2,
+                        0b11 => 4,
+                        _ => unreachable!(), // Contradicts match on last 3 bits.
+                    };
+                ud.payload.resize_default(payload_size)
+                    .expect("payload_size <= payload.capacity");
+                match self.inner.read_exact(&mut *ud.payload) {
+                    Err(ref e) if e.kind() == io::ErrorKind::UnexpectedEof =>
+                        return Err(Error::from(ErrorKind::EofDuringPacket)),
+                    Err(e) => return Err(Error::from(e)),
+                    Ok(_) => (),
+                };
+
+                Ok(Packet {
+                    header: header,
+                    kind: packet::Kind::Instrumentation(ud),
+                })
+            },
+            _ => {
+                return Err(Error::from(ErrorKind::UnknownHeader(header)));
+            }
+        }
+    }
+}
+
+/// Decode a single packet from a slice of bytes.
+pub fn from_slice(s: &[u8]) -> Result<Packet> {
```

Comment:
Not really important but is it possible to return a `Result` type here that doesn't include the `Io` variant? I think that no I/O error can occur in this function so it would be nice to reflect that in the signature. I'm not sure if error-chain allows such thing though; I have never used the error-chain! macro with the `errors` fields.

Reply:
Won't change, as discussed.